design/ceph/osd-migration.md

```diff
@@ -0,0 +1,207 @@
+	LastReplacedOSD      *OSDReplaceInfo `json:"lastReplacedOSD,omitempty"`
+	LastReplaceTimeStamp time.Time       `json:"lastReplacedTimeStamp,omitempty"`
+	StartTimeStamp       time.Time       `json:"startTimeStamp,omitempty"`
+}
+```
+*OSDReplaceInfo*: Represents an OSD that needs to be replace
+  - *ID*: OSD ID of the OSD to be replace.
+  - *Path*: Disk or pvc claim that the OSD is consuming.
+  - *Node*: Node on which the OSD is running
+
+
+*ReplaceQueue*: This is a queue implementation containing a list of OSDs to be replaced. It will support basic queue operations like pop, push, size etc.
+
+
+*OSDReplaceConfig*: This will hold the current status/progress of the OSD replacement in the cluster.
+  - *OSDList*:  List of OSDs that are yet to be replaced.
+  - *IsPgHealth*: Whether Placement Groups are healthy or not. This is a boolean field.
+  - *PgHealthMessage*: This will contain a detailed health message of the placement groups. For example: `active+clean`
+  - *LastReplaceOSD*: This field will contain details of the OSD that was last replaced.
+  - *LastReplaceTimeStamp*: Time at which the last OSD was replaced.
+  - *StartTimeStamp*: Time at which the whole OSD replacement process was started. This will help users know how much time has elapsed since the first OSD was replaced.
+
+
+#### Replace OSD :
+Operator reconciler will perform following sequence of steps to replace an OSD:
+  1. List out all the OSDs that need to be replaced. This list will contain OSDs that have `osd-store:<osd store type>` pod label not matching the osd store type provided in `spec.storage.osdStore.type`. This list will be stored in `OSDReplaceConfig.OSDList` and will be used a queue.
+  2. If the PG status is not `active+clean`, don't pop and OSD from the `OSDReplaceConfig.OSDList`
+  3. Once the PG status is `active+clean`, pop an OSD entry from the `OSDReplaceConfig.OSDList`. Update the `LastReplacedOSD` and `LastReplaceTimeStamp`
+  4. Provide two new environment variables to the OSD prepare pod running on `OSDReplaceInfo.Node` in case of OSD on disks or `OSDReplaceInfo.Path` in case of OSD on PVCs
+   - `ROOK_REPLACE_OSD_ID`: OSD to be deleted.
+   - `ROOK_REPLACE_OSD_PATH`: The backing device or pvc claim of the OSD
+  5. OSD prepare pod job will destroy the OSD based on the environment variables provided above. This process is mentioned in the `Destroy OSD`` steps below.
+  6. Continue with the existing flow of creating and upgrading the OSD, with the exception of skipping the upgrade for all the OSDs that are currently present in the `OSDReplaceConfig.OSDList`. Skipping update operation on the OSDs that need to be migrated is important because with OSDs on disks, the OSD prepared pod is updated if there is an update in the spec. This will update the osd store to `bluestore-rdr` in the spec without migrating the OSD. This will cause errors.
+  7. Reconcile the operator and perform as same steps in the next reconcile.
+
+
+
```

Comment:
Nit to make spacing consistent between sections. There are 3 here but it looks like 2 elsewhere.

Reply:
Made it consistent with some other docs by using only single line spacing for all the section.